lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,193 @@
+        end
+      end
+
+      def trace(key, data, &block)
+        skip_tracing = data[:query]&.context&.fetch(:skip_graphql_metrics_analysis, false)
+        return block.call if skip_tracing || ![GRAPHQL_TRACING_FIELD_KEY, GRAPHQL_TRACING_LAZY_FIELD_KEY].include?(key)
+        # NOTE: We can't just check `runtime_metrics_enabled?` here, since .trace runs before `before_query`, i.e.
+        # during lexing, parsing, validation etc., and `before_query` doesn't run until `execute_multiplex`.
+
+        if runtime_metrics_enabled?(data[:query]&.context)
+          trace_field(key, data, block, key)
+        else
+          block.call
+        end
+      end
+
+      def trace_field(_key, data, block, context_key)
+        path_excluding_numeric_indicies = data[:path].select { |p| p.is_a?(String) }
+
+        start_time = current_time
+        start_time_monotonic = current_time_monotonic
+
+        result = block.call
+        duration = current_time_monotonic - start_time_monotonic
+
+        data[:query].context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[context_key][path_excluding_numeric_indicies] ||= []
+          ns[context_key][path_excluding_numeric_indicies] << { start_time: start_time, duration: duration }
+        end
+
+        result
+      end
+
+      def runtime_metrics_enabled?(context)
+        return false unless context
+        !!context.namespace(CONTEXT_NAMESPACE)[RUNTIME_METRICS_ENABLED]
+      end
+    end
+
+    def initialize(query_or_multiplex)
+      super
+
+      query_or_multiplex.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+        ns[ANALYZER_INSTANCE_KEY] = self
+      end
+
+      @query = query_or_multiplex
+      @static_query_metrics = nil
+      @field_timing_callbacks = []
+    end
+
+    def analyze?
+      query.valid? && query.context[:skip_graphql_metrics_analysis] != true
+    end
+
+    def extract_query(runtime_metrics: {}, context:)
+      query_extracted(@static_query_metrics.merge(runtime_metrics))
+    end
+
+    # TODO: Apollo Tracing spec https://docs.google.com/document/d/1B0dR09CcN_M4yqezkJ7VFPI-mKgQIVCt9PiUcIMEVNI/edit
+    # Do so after integrating with Shopify/shopify, since we'll need another Instrumentation class to dig out values
+    # left in memory by Analyzer.
+
+    def on_enter_operation_definition(_node, _parent, visitor)
+      @static_query_metrics = {
+        operation_type: visitor.query.selected_operation.operation_type,
+        operation_name: visitor.query.selected_operation.name,
+      }
+    end
+
+    def on_leave_field(node, _parent, visitor)
+      static_metrics = {
+        field_name: node.name,
+        return_type_name: visitor.type_definition.name,
+        parent_type_name: visitor.parent_type_definition.name,
+        deprecated: visitor.field_definition.deprecation_reason.present?,
+        path: visitor.response_path,
+      }
+
+      if self.class.runtime_metrics_enabled?(visitor.query.context)
+        callback = -> (metrics) {
```

Comment:
perhaps a micro-optimization, but right now we're recreating this callback on every field, where I think it's going to be identical for every one? the only out-of-closure reference is to `visitor` which I *think* is constant across the whole query?

Reply:
Good point! I don't think this was always the case before recent changes. I'll revisit this decision to store a callback per field visited.